network/p2p/dht.go

```diff
@@ -0,0 +1,59 @@
+package p2p
+
+import (
+	"context"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	discovery "github.com/libp2p/go-libp2p-discovery"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+)
+
+// This produces a new IPFS DHT
+// on the name, see https://github.com/libp2p/go-libp2p-kad-dht/issues/337
+func newDHT(ctx context.Context, host host.Host, serverMode bool) (*discovery.RoutingDiscovery, error) {
+	defaultOptions := defaultDHTOptions()
+
+	if serverMode {
+		// DHT defaults to ModeAuto which will automatically switch the DHT between Server and Client modes based on
+		// whether the node appears to be publicly reachable (e.g. not behind a NAT and with a public IP address).
+		// This default tends to make test setups fail (since the test nodes are normally not reachable by the public
+		// network), but is useful for improving the stability and performance of live public networks.
+		// While we could force all nodes to be DHT Servers, a bunch of nodes otherwise not reachable by most of the
+		// network => network partition
+		defaultOptions = append(defaultOptions, dht.Mode(dht.ModeServer))
+	}
+
+	kdht, err := dht.New(ctx, host, defaultOptions...)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = kdht.Bootstrap(ctx); err != nil {
+		return nil, err
+	}
+
+	routingDiscovery := discovery.NewRoutingDiscovery(kdht)
+	return routingDiscovery, nil
+}
+
+// NewDHTServer starts the DHT in server mode
+func NewDHTServer(ctx context.Context, host host.Host) (*discovery.RoutingDiscovery, error) {
+	return newDHT(ctx, host, true)
+}
+
+// NewDHTClient starts the DHT in client mode
+func NewDHTClient(ctx context.Context, host host.Host) (*discovery.RoutingDiscovery, error) {
+	return newDHT(ctx, host, false)
+}
+
+func defaultDHTOptions() []dht.Option {
+	return []dht.Option{
+
+		dht.ProtocolPrefix(FlowLibP2PProtocolCommonPrefix),
+
+		// public capabilities we don't need - disabling these capabilities does not allow peer discovery leading
```

Comment:
My understanding was that only staked nodes would actually act as DHT servers? Since `RoutingTableRefreshPeriod` means how often we will refresh the routing table, and only staked nodes are part of the routing table, we don't really consider them ephemeral nodes right?

Reply:
Also, if we ever fail to connect to a peer in our routing table they will end up getting dropped from the routing table anyways, so I'm not sure we need to explicitly refresh so often?